### PR TITLE
osd: object added to missing set for backfill, but is not in recovering, error!

### PR DESCRIPTION
**Pull request by dzafman**

### CODE REVIEW

**Comment by dzafman** [comment]
@jdurgin The test confirms that the _rados get_ hangs for the object that is marked lost.  After _mark_unfound_lost delete_, the _rados get_ returns _ENOENT_.

**Comment by dzafman** [comment]
Standalone tests:

http://pulpito.ceph.com/dzafman-2017-10-12_18:54:04-rados:standalone-wip-zafman-testing-distro-basic-smithi/

**Comment by dzafman** [comment]
All but 5 tests passed.  The 5 failures seem unrelated and the re-run of those jobs all passed!
http://pulpito.ceph.com/dzafman-2017-10-12_20:04:43-rados-wip-zafman-testing-distro-basic-smithi

 http://pulpito.ceph.com/dzafman-2017-10-13_10:39:12-rados-wip-zafman-testing-distro-basic-smithi

File: qa/standalone/erasure-code/test-erasure-eio.sh [hunk 1]
```diff
@@ -403,7 +377,159 @@ function TEST_ec_recovery_errors() {
...
+    wait_for_clean || return 1
+
+    for i in $(seq 1 $lastobj)
+    do
+      if [ obj${i} = "$testobj" ]; then
+        # Doesn't exist anymore
+        ! rados -p $poolname get $testobj $dir/CHECK || return 1
+      else
+        rados --pool $poolname get obj${i} $dir/CHECK || return 1
+        diff -q $dir/ORIGINAL $dir/CHECK | return 1
```
Comment (jdurgin) [inline qa/standalone/erasure-code/test-erasure-eio.sh]: typo: | should be ||

File: qa/standalone/erasure-code/test-erasure-eio.sh [hunk 2]
```diff
@@ -403,7 +377,159 @@ function TEST_ec_recovery_errors() {
...
```
Comment (jdurgin) [inline qa/standalone/erasure-code/test-erasure-eio.sh]: same typo

File: qa/standalone/osd/osd-rep-recov-eio.sh [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    wait_for_clean || return 1
+
+    for i in $(seq 1 $lastobj)
+    do
+      if [ obj${i} = "$testobj" ]; then
+        # Doesn't exist anymore
+        ! rados -p $poolname get $testobj $dir/CHECK || return 1
+      else
+        rados --pool $poolname get obj${i} $dir/CHECK || return 1
+        diff -q $dir/ORIGINAL $dir/CHECK | return 1
```
Comment (jdurgin) [inline qa/standalone/osd/osd-rep-recov-eio.sh]: typo here

File: qa/standalone/osd/osd-rep-recov-eio.sh [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
```
Comment (jdurgin) [inline qa/standalone/osd/osd-rep-recov-eio.sh]: and here

File: src/osd/PrimaryLogPG.cc [hunk 5]
```diff
@@ -10572,7 +10581,24 @@ void PrimaryLogPG::mark_all_unfound_lost(
...
+	      DoRecovery())));
+	} else if (is_backfill_unfound()) {
+	  queue_peering_event(
+	    CephPeeringEvtRef(
+	      std::make_shared<CephPeeringEvt>(
+	      get_osdmap()->get_epoch(),
+	      get_osdmap()->get_epoch(),
+	      RequestBackfill())));
+	} else {
+	  // Not sure if this is needed anymore
```
Comment (jdurgin) [inline src/osd/PrimaryLogPG.cc]: I guess we could replace it with an assert to catch bugs where we're not in {backfill,recovery}_unfound

**Pull request merged**